File: luaPatchDemo/ApplePay.m
//
//  ApplePay.m
//  luaPatchDemo
//
//  Created by 黄钊 on 2019/8/23.
//  Copyright © 2019 hz. All rights reserved.
//

#import "ApplePay.h"

@implementation ApplePay

//必须要有init函数，否则SKPaymentQueue defaultQueue不能反射调用成功，原因可能是因为编译器在代码中没有找到SKPaymentQueue符号的话就不会把支付模块编译进去
-(instancetype)init{
    if(self = [super init]){
        [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
        
//        Class cls = NSClassFromString(@"SKPaymentQueue");
//        NSInvocation *invocation;
//        NSMethodSignature *methodSignature = [cls methodSignatureForSelector:@selector(defaultQueue)];
//        invocation = [NSInvocation invocationWithMethodSignature:methodSignature];
//        [invocation setSelector:@selector(defaultQueue)];
//        [invocation invoke];
        
        return self;
    }
    return nil;
}


- (void)paymentQueue:(nonnull SKPaymentQueue *)queue updatedTransactions:(nonnull NSArray<SKPaymentTransaction *> *)transactions {
    for(SKPaymentTransaction *transaction in transactions){
        switch (transaction.transactionState) {
            case SKPaymentTransactionStatePurchased:
                NSLog(@"paymentQueue SKPaymentTransactionStatePurchased");
                [[SKPaymentQueue defaultQueue]finishTransaction:transaction];
                break;
                
            case SKPaymentTransactionStateFailed:
                NSLog(@"paymentQueue SKPaymentTransactionStateFailed");
                [[SKPaymentQueue defaultQueue]finishTransaction:transaction];
                break;
            
            default:
                NSLog(@"paymentQueue default %li",(long)transaction.transactionState);
                break;
        }
    }
}

@end

File: luaSource/ApplePay.lua
local luapatch = require("luapatch")
luapatch.redefineInstanceMethod("ApplePay","init","ApplePay_init","id")
--luapatch.addInstanceMethod("ApplePay","init","ApplePay_init","void")
function ApplePay_init(instance)
    -- body
    instance.super().init()
    luapatch.packClass("SKPaymentQueue").defaultQueue().addTransactionObserver(instance)
    return instance
end

luapatch.addInstanceMethod("ApplePay","pay:","ApplePay_openpay","void_id")
function ApplePay_openpay(instance,productIdentifier)
    -- body
    local SKMutablePayment = luapatch.packClass("SKMutablePayment")
    local mutablePayment = SKMutablePayment.alloc().init()
    mutablePayment.setProductIdentifier(productIdentifier)
    mutablePayment.setQuantity(1)
    luapatch.packClass("SKPaymentQueue").defaultQueue().addPayment(mutablePayment)
end

luapatch.redefineInstanceMethod("ApplePay","paymentQueue:updatedTransactions:","ApplePay_paymentQueue_updatedTransactions","void_id_id")
function ApplePay_paymentQueue_updatedTransactions(instance,queue,transactions)
    -- body
    print("call ApplePay_paymentQueue_updatedTransactions")
    local count = transactions.count()
    print("count is ")
    print(count)
    local transaction = {}
    for i = 0, count-1 do
        transaction = transactions.objectAtIndex(i)
        if transaction.transactionState() == 1 then --SKPaymentTransactionStatePurchased
            print("transaction is SKPaymentTransactionStatePurchased in lua")
            luapatch.packClass("SKPaymentQueue").defaultQueue().finishTransaction(transaction)
        elseif transaction.transactionState() == 2 then --SKPaymentTransactionStateFailed
            print("transaction is SKPaymentTransactionStateFailed in lua")
            luapatch.packClass("SKPaymentQueue").defaultQueue().finishTransaction(transaction)
        else
            print("transaction is default in lua")
        end
    end
end
